core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/guess.kt

```diff
@@ -16,18 +19,47 @@ import java.net.URL
 import java.util.ServiceLoader
 import kotlin.reflect.KType
 
-public interface SupportedFormat {
-    public fun readDataFrame(stream: InputStream, header: List<String> = emptyList()): AnyFrame
-
-    public fun readDataFrame(file: File, header: List<String> = emptyList()): AnyFrame
+public sealed interface SupportedFormat {
 
     public fun acceptsExtension(ext: String): Boolean
 
     // `DataFrame.Companion.read` methods uses this to sort list of all supported formats in ascending order (-1, 2, 10)
     // sorted list is used to test if any format can read given input
     public val testOrder: Int
 
-    public fun createDefaultReadMethod(pathRepresentation: String?): DefaultReadDfMethod
+    public fun createDefaultReadMethod(pathRepresentation: String? = null): DefaultReadDfMethod
+}
+
+/**
+ * Implement this interface to provide additional supported formats for DataFrames (such as JSON, XML, CSV, etc.).
+ * A [SupportedDataFrameFormat] is read directly to a DataFrame. If specified using
+ * [ImportDataSchema] or using the Gradle plugin, the read DataFrame will be used to
+ * generate [DataSchema] interfaces.
+ */
+public interface SupportedDataFrameFormat : SupportedFormat {
+    public fun readDataFrame(stream: InputStream, header: List<String> = emptyList()): DataFrame<*>
+
+    public fun readDataFrame(file: File, header: List<String> = emptyList()): DataFrame<*>
+}
+
+/**
+ * Implement this interface to provide additional [DataSchema] interface generation formats for DataFrames (such as OpenAPI).
+ * Note, this doesn't add functionality to [DataFrame.Companion.read], just [ImportDataSchema] and Gradle plugin.
+ *
+ * Return type will be a [CodeWithConverter] for which [CodeWithConverter.converter] needs to be called
+ * to get the generated code with the correct top-level interface name.
+ */
+public interface SupportedCodeGenerationFormat : SupportedFormat {
+
+    public fun readCodeForGeneration(
+        stream: InputStream,
+        generateHelperCompanionObject: Boolean = false,
```

Comment:
this in fact needed to support val ApiGuru = importDataSchema() notation in Jupyter, right? I think it's worth mentioning

Reply:
yup!